File: frontend/lib/ghost.ts
/**
 * Aethel Ghost UI
 * Pre-Cognitive Execution Interface
 * 
 * "The answer exists before the question is complete."
 */

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';
const LATTICE_NODES_RAW = process.env.NEXT_PUBLIC_LATTICE_NODES || '';

function getCandidateBaseUrls(): string[] {
  const nodes = LATTICE_NODES_RAW
    .split(',')
    .map((s) => s.trim())
    .filter(Boolean)
    .map((s) => s.replace(/\/$/, ''));

  const primary = API_URL.replace(/\/$/, '');

  const seen = new Set<string>();
  const out: string[] = [];

  for (const u of [primary, ...nodes]) {
    if (!seen.has(u)) {
      seen.add(u);
      out.push(u);
    }
  }

  return out;
}

async function fetchWithFallback(path: string, init: RequestInit): Promise<Response> {
  const bases = getCandidateBaseUrls();
  let lastError: unknown = null;

  for (const base of bases) {
    try {
      const resp = await fetch(`${base}${path}`, init);
      if (resp.ok) return resp;
      lastError = new Error(`HTTP error! status: ${resp.status}`);
    } catch (e) {
      lastError = e;
    }
  }

  throw lastError instanceof Error ? lastError : new Error('All lattice nodes failed');
}

export interface GhostState {
  variables: Record<string, any> | null;
  merkle_root: string | null;
}

export interface GhostPrediction {
  success: boolean;
  status: 'MANIFESTED' | 'IMPOSSIBLE' | 'UNCERTAIN' | 'ERROR';
  confidence: number;
  latency: number;
  eliminated_states: number;
  message: string;
  result: GhostState | null;
}

export class GhostUI {
  private debounceTimer: NodeJS.Timeout | null = null;
  private cache: Map<string, GhostPrediction> = new Map();
  
  /**
   * Manifests truth while user types.
   * Doesn't wait for click - truth already exists.
   */
  async manifestTruth(code: string): Promise<GhostPrediction> {
    // Check cache first - truth is eternal
    const cacheKey = this.hashCode(code);
    if (this.cache.has(cacheKey)) {
      return this.cache.get(cacheKey)!;
    }
    
    // Debounce to avoid overwhelming the server
    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer);
    }
    
    return new Promise((resolve) => {
      this.debounceTimer = setTimeout(async () => {
        try {
          const response = await fetchWithFallback(`/api/ghost/predict`, {
            method: 'POST',
            headers: {
              'Content-Type': 'application/json',
            },
            body: JSON.stringify({ code }),
          });
          
          const prediction: GhostPrediction = await response.json();
          
          // Cache the manifestation
          if (prediction.status === 'MANIFESTED') {
            this.cache.set(cacheKey, prediction);
          }
          
          resolve(prediction);
          
        } catch (error) {
          resolve({
            success: false,
            status: 'ERROR',
            confidence: 0,
            latency: 0,
            eliminated_states: 0,
            message: `Ghost-Runner error: ${error}`,
            result: null,
          });
        }
      }, 500); // 500ms debounce
    });
  }
  
  /**
   * Prevents typing impossible code.
   * Returns false if next character leads to impossible state.
   * 
   * This is the "cursor lock" feature.
   */
  async canTypeNextChar(code: string, nextChar: string): Promise<boolean> {
    try {
      const response = await fetchWithFallback(`/api/ghost/can-type`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ 
          code, 
          nextChar 
        }),
      });
      
      const result = await response.json();
      return result.canType !== false; // Default to true if uncertain
      
    } catch {
      return true; // Fail open - allow typing if check fails
    }
  }
  
  /**
   * Gets confidence level as percentage
   */
  getConfidencePercent(prediction: GhostPrediction): number {
    return Math.round(prediction.confidence * 100);
  }
  
  /**
   * Gets status color for UI
   */
  getStatusColor(status: string): string {
    switch (status) {
      case 'MANIFESTED':
        return 'text-green-400';
      case 'IMPOSSIBLE':
        return 'text-red-400';
      case 'UNCERTAIN':
        return 'text-yellow-400';
      default:
        return 'text-gray-400';
    }
  }
  
  /**
   * Gets status icon
   */
  getStatusIcon(status: string): string {
    switch (status) {
      case 'MANIFESTED':
        return '✨';
      case 'IMPOSSIBLE':
        return '🚫';
      case 'UNCERTAIN':
        return '🔮';
      default:
        return '❓';
    }
  }
  
  private hashCode(str: string): string {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash;
    }
    return hash.toString(36);
  }
}

// Singleton instance
let ghostInstance: GhostUI | null = null;

export function getGhostUI(): GhostUI {
  if (!ghostInstance) {
    ghostInstance = new GhostUI();
  }
  return ghostInstance;
}

File: huggingface_deploy_package/diotec360/core/ghost.py
"""
Copyright 2024 Dionísio Sebastião Barros / DIOTEC 360

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

"""
Aethel Ghost-Runner
The Secret of the Sand: Subtraction of the Impossible

This module implements Zero-Latency Computing through
Pre-Cognitive Execution. Instead of calculating results,
it manifests truth by eliminating impossible states.

"The answer exists before the question is complete."
"""

from typing import Dict, List, Any, Optional, Set
from dataclasses import dataclass
import hashlib
import json


@dataclass
class GhostState:
    """
    Represents a possible state in the universe of computation.
    """
    variables: Dict[str, Any]
    merkle_root: str
    confidence: float  # 1.0 = mathematically certain
    
    def __hash__(self):
        return hash(self.merkle_root)


@dataclass
class GhostPrediction:
    """
    The manifestation of truth - result before execution.
    """
    status: str  # 'MANIFESTED', 'IMPOSSIBLE', 'UNCERTAIN'
    result: Optional[GhostState]
    confidence: float
    latency: float  # Always 0 for manifested states
    eliminated_states: int  # How many impossibilities were subtracted
    message: str


class GhostRunner:
    """
    The Ghost-Runner doesn't execute code.
    It manifests truth by subtracting the impossible.
    
    Core Principle:
    - Traditional: Build the answer step by step
    - Ghost: Remove all wrong answers, what remains IS the answer
    """
    
    def __init__(self, judge=None, state_manager=None):
        self.judge = judge
        self.state = state_manager
        self.cache = {}  # Cache of manifested truths
        
    def predict_outcome(self, intent_ast: Dict) -> GhostPrediction:
        """
        Predicts the outcome BEFORE execution.
        
        How? The Judge already proved only ONE state is valid.
        We don't need to calculate - just manifest.
        
        Args:
            intent_ast: Parsed Aethel intent
            
        Returns:
            GhostPrediction with the manifested truth
        """
        
        # Check cache first - truth is eternal
        cache_key = self._hash_intent(intent_ast)
        if cache_key in self.cache:
            cached = self.cache[cache_key]
            return GhostPrediction(
                status='MANIFESTED',
                result=cached,
                confidence=1.0,
                latency=0.0,
                eliminated_states=0,
                message='Truth retrieved from eternal cache'
            )
        
        try:
            # 1. Extract constraints from intent
            guards = self._extract_guards(intent_ast)
            verifications = self._extract_verifications(intent_ast)
            
            # 2. Generate universe of possible states
            all_states = self._generate_state_space(intent_ast)
            initial_count = len(all_states)
            
            # 3. SUBTRACTION: Eliminate impossible states
            valid_states = set(all_states)
            
            # Subtract states that violate guards
            for guard in guards:
                valid_states = self._subtract_invalid(valid_states, guard)
            
            # Subtract states that violate verifications
            for verify in verifications:
                valid_states = self._subtract_invalid(valid_states, verify)
            
            eliminated = initial_count - len(valid_states)
            
            # 4. What remains IS the truth
            if len(valid_states) == 1:
                # Perfect manifestation - only one truth exists
                truth = list(valid_states)[0]
                self.cache[cache_key] = truth
                
                return GhostPrediction(
                    status='MANIFESTED',
                    result=truth,
                    confidence=1.0,
                    latency=0.0,
                    eliminated_states=eliminated,
                    message=f'Truth manifested by eliminating {eliminated} impossible states'
                )
                
            elif len(valid_states) == 0:
                # Impossible - no valid state exists
                return GhostPrediction(
                    status='IMPOSSIBLE',
                    result=None,
                    confidence=0.0,
                    latency=0.0,
                    eliminated_states=initial_count,
                    message='All states eliminated - this intent is impossible'
                )
                
            else:
                # Multiple valid states - need more constraints
                # Choose the first one but mark as uncertain
                truth = list(valid_states)[0]
                
                return GhostPrediction(
                    status='UNCERTAIN',
                    result=truth,
                    confidence=1.0 / len(valid_states),
                    latency=0.0,
                    eliminated_states=eliminated,
                    message=f'{len(valid_states)} possible truths remain - need more constraints'
                )
                
        except Exception as e:
            return GhostPrediction(
                status='ERROR',
                result=None,
                confidence=0.0,
                latency=0.0,
                eliminated_states=0,
                message=f'Ghost-Runner error: {str(e)}'
            )
    
    def can_type_next_char(self, current_code: str, next_char: str) -> bool:
        """
        Prevents typing impossible code.
        Returns False if next_char would lead to an impossible state.
        
        This is the "cursor lock" feature - the keyboard physically
        prevents you from typing bugs.
        """
        future_code = current_code + next_char
        
        try:
            # Quick syntax check
            if not self._is_valid_syntax(future_code):
                return False
            
            # If it's a complete intent, check if it's possible
            if self._is_complete_intent(future_code):
                ast = self._parse_code(future_code)
                prediction = self.predict_outcome(ast)
                return prediction.status != 'IMPOSSIBLE'
            
            # Partial code is always allowed
            return True
            
        except:
            # If we can't determine, allow it (fail open)
            return True
    
    def _generate_state_space(self, intent_ast: Dict) -> List[GhostState]:
        """
        Generates the universe of all possible states.
        
        In practice, this is bounded by the variables in the intent.
        For a simple transfer, this might be:
        - sender_balance: [0, 100, 1000, 10000]
        - receiver_balance: [0, 100, 1000, 10000]
        - amount: [0, 50, 100, 500]
        
        Total: 4 * 4 * 4 = 64 possible states
        """
        
        # Extract variables from intent
        variables = self._extract_variables(intent_ast)
        
        # Generate sample values for each variable
        # In production, this would be smarter
        states = []
        
        # For now, generate a small sample space
        sample_values = [0, 10, 100, 1000]
        
        # Generate all combinations (simplified)
        for val1 in sample_values:
            for val2 in sample_values:
                state = GhostState(
                    variables={
                        'sender_balance': val1,
                        'receiver_balance': val2,
                        'amount': min(val1, val2)
                    },
                    merkle_root=self._compute_merkle_root({
                        'sender_balance': val1,
                        'receiver_balance': val2
                    }),
                    confidence=1.0
                )
                states.append(state)
        
        return states
    
    def _subtract_invalid(self, states: Set[GhostState], constraint: str) -> Set[GhostState]:
        """
        Removes states that violate the constraint.
        
        This is the core of the Ghost-Runner:
        We don't build the answer - we eliminate wrong answers.
        """
        valid_states = set()
        
        for state in states:
            if self._check_constraint(state, constraint):
                valid_states.add(state)
        
        return valid_states
    
    def _check_constraint(self, state: GhostState, constraint: str) -> bool:
        """
        Checks if a state satisfies a constraint.
        
        This is where the Judge (Z3) would be called in production.
        For now, we do simple evaluation.
        """
        try:
            # Simple constraint evaluation
            # In production, this would use Z3
            
            # Example: "sender_balance >= amount"
            if '>=' in constraint:
                parts = constraint.split('>=')
                left = parts[0].strip()
                right = parts[1].strip()
                
                left_val = state.variables.get(left, 0)
                right_val = state.variables.get(right, 0) if right in state.variables else int(right)
                
                return left_val >= right_val
            
            # Add more constraint types as needed
            return True
            
        except:
            return True  # Fail open
    
    def _extract_guards(self, intent_ast: Dict) -> List[str]:
        """Extract guard constraints from intent AST"""
        guards = intent_ast.get('guards', [])
        return [g.get('condition', '') for g in guards]
    
    def _extract_verifications(self, intent_ast: Dict) -> List[str]:
        """Extract verification constraints from intent AST"""
        verifications = intent_ast.get('verifications', [])
        return [v.get('condition', '') for v in verifications]
    
    def _extract_variables(self, intent_ast: Dict) -> List[str]:
        """Extract variable names from intent AST"""
        # Simplified - in production would parse AST properly
        return ['sender_balance', 'receiver_balance', 'amount']
    
    def _compute_merkle_root(self, variables: Dict) -> str:
        """Compute Merkle root for a state"""
        data = json.dumps(variables, sort_keys=True)
        return hashlib.sha256(data.encode()).hexdigest()
    
    def _hash_intent(self, intent_ast: Dict) -> str:
        """Hash an intent for caching"""
        data = json.dumps(intent_ast, sort_keys=True)
        return hashlib.sha256(data.encode()).hexdigest()
    
    def _is_valid_syntax(self, code: str) -> bool:
        """Quick syntax validation"""
        # Simplified - just check for basic structure
        return True
    
    def _is_complete_intent(self, code: str) -> bool:
        """Check if code is a complete intent"""
        return 'intent' in code and '{' in code and '}' in code
    
    def _parse_code(self, code: str) -> Dict:
        """Parse Aethel code to AST"""
        # Simplified - in production would use real parser
        return {
            'name': 'transfer',
            'guards': [],
            'verifications': []
        }


# Singleton instance
_ghost_runner = None

def get_ghost_runner():
    """Get the global Ghost-Runner instance"""
    global _ghost_runner
    if _ghost_runner is None:
        _ghost_runner = GhostRunner()
    return _ghost_runner
